internal/querynode/distribution.go

```diff
@@ -160,76 +162,46 @@ func (d *distribution) NodeDown(nodeID int64) {
 	d.mut.Lock()
 	defer d.mut.Unlock()
 
-	var delta int32
-
+	d.getLogger().Info("handle node down", zap.Int64("node", nodeID))
 	for _, entry := range d.sealedSegments {
 		if entry.NodeID == nodeID && entry.State != segmentStateOffline {
 			entry.State = segmentStateOffline
 			d.sealedSegments[entry.SegmentID] = entry
-			delta++
+			d.getLogger().Info("update the segment to offline since nodeDown", zap.Int64("nodeID", nodeID), zap.Int64("segmentID", entry.SegmentID))
 		}
 	}
-
-	if delta != 0 {
-		d.offlines.Add(delta)
-		d.getLogger().Info("distribution updated since nodeDown", zap.Int32("delta", delta), zap.Int32("offlines", d.offlines.Load()), zap.Int64("nodeID", nodeID))
-	}
-}
-
-// updateSegment update segment entry value and offline segment number based on old/new state.
-func (d *distribution) updateSegment(old, new SegmentEntry) {
-	delta := int32(0)
-	switch {
-	case old.State != segmentStateLoaded && new.State == segmentStateLoaded:
-		delta = -1
-	case old.State == segmentStateLoaded && new.State != segmentStateLoaded:
-		delta = 1
-	}
-
-	old.Update(new)
-	d.sealedSegments[old.SegmentID] = old
-	if delta != 0 {
-		d.offlines.Add(delta)
-		d.getLogger().Info("distribution updated since segment update",
-			zap.Int32("delta", delta),
-			zap.Int32("offlines", d.offlines.Load()),
-			zap.Int64("segmentID", new.SegmentID),
-			zap.Int32("state", int32(new.State)),
-		)
-	}
 }
 
 // RemoveDistributions remove segments distributions and returns the clear signal channel,
 // requires the read lock of shard cluster mut held
 func (d *distribution) RemoveDistributions(releaseFn func(), sealedSegments ...SegmentEntry) {
 	d.mut.Lock()
-	defer d.mut.Unlock()
 
-	var delta int32
 	for _, sealed := range sealedSegments {
 		entry, ok := d.sealedSegments[sealed.SegmentID]
+		d.getLogger().Info("Remove distribution", zap.Int64("segmentID", sealed.SegmentID),
+			zap.Int64("node", sealed.NodeID),
+			zap.Bool("segment exist", ok))
 		if !ok {
 			continue
 		}
 		if entry.NodeID == sealed.NodeID || sealed.NodeID == wildcardNodeID {
-			if entry.State == segmentStateOffline {
-				delta--
-			}
 			delete(d.sealedSegments, sealed.SegmentID)
 		}
 	}
 
-	d.offlines.Add(delta)
+	// we do not want to hold mut lock while we wait for search.
+	sig := d.genSnapshot()
+	d.mut.Unlock()
 
-	<-d.genSnapshot()
+	<-sig
 	releaseFn()
```

Comment:
it's designed to hold the lock here since this `releaseFn` may release some growing segments here. 
If the lock is released before executing the `releaseFn` some future search/query may access this growing segment and cause panicking